prog/vm/github_runner.rb

```diff
@@ -155,50 +155,50 @@ def before_run
-          fail "BUG: Failed with runner already exists error but couldn't find it"
-        end
-        Clog.emit("Deleting GithubRunner because it already exists") { {github_runner: github_runner.values.merge({runner_id: runner[:id]})} }
-        github_client.delete("/repos/#{github_runner.repository_name}/actions/runners/#{runner[:id]}")
-        nap 5
-      end
-
-      github_runner.update(runner_id: response[:runner][:id], ready_at: Time.now)
-
-      command = "./actions-runner/run.sh --jitconfig #{response[:encoded_jit_config].shellescape}"
-      vm.sshable.cmd("sudo systemd-run --uid runner --gid runner --working-directory '/home/runner' --unit #{SERVICE_NAME} --remain-after-exit -- #{command}")
+    # We use generate-jitconfig instead of registration-token because it's
+    # recommended by GitHub for security reasons.
+    # https://docs.github.com/en/actions/security-guides/security-hardening-for-github-actions#using-just-in-time-runners
+    data = {name: github_runner.ubid.to_s, labels: [github_runner.label], runner_group_id: 1}
+    response = github_client.post("/repos/#{github_runner.repository_name}/actions/runners/generate-jitconfig", data)
+    github_runner.update(runner_id: response[:runner][:id], ready_at: Time.now)
+
+    # We initiate an API call and a SSH connection under the same label to avoid
+    # having to store the encoded_jit_config.
+    command = "./actions-runner/run.sh --jitconfig #{response[:encoded_jit_config].shellescape}"
+    vm.sshable.cmd("sudo systemd-run --uid runner --gid runner --working-directory '/home/runner' --unit #{SERVICE_NAME} --remain-after-exit -- #{command}")
+
+    hop_wait
+  rescue Octokit::Conflict => e
+    unless e.message.include?("Already exists")
+      raise e
+    end
+    # If the runner already exists and the model lacks a 'runner_id'
+    # column, this suggests that the process terminated prematurely before
+    # the 'runner_id' column could be updated. We need to locate the
+    # 'runner_id' using the name and delete it.
+    # After this, we can register the runner again.
+    runners = github_client.paginate("/repos/#{github_runner.repository_name}/actions/runners") do |data, last_response|
+      data[:runners].concat last_response.data[:runners]
     end
+    unless (runner = runners[:runners].find { _1[:name] == github_runner.ubid.to_s })
+      fail "BUG: Failed with runner already exists error but couldn't find it"
+    end
+    Clog.emit("Deleting GithubRunner because it already exists") { {github_runner: github_runner.values.merge({runner_id: runner[:id]})} }
+    github_client.delete("/repos/#{github_runner.repository_name}/actions/runners/#{runner[:id]}")
+    nap 5
+  end
 
+  label def wait
     case vm.sshable.cmd("systemctl show -p SubState --value #{SERVICE_NAME}").chomp
-    when "exited", "running"
-      hop_wait
+    when "exited"
+      github_runner.incr_destroy
```

Comment:
why don't you `hop_destroy`?

Reply:
"before_run" might have a logic. For example it finalizes the billing records at vm nexus.

Reply:
This is getting out of hand, we should keep that logic consistent. For now, you don't have anything in `before_run`, it just allows us to destroy from any state. So, maybe, we should not allow `hop_destroy` ever.

I'm good with your change here.

Reply:
I've reviewed the codebase, and all calls to `hop_destroy` are located within the `before_run` function, with the exception of `Prog::Vnet::SubnetNexus`. I need to further consider how we can restrict the use of `hop_destroy` to only the `before_run` function.